File: functions/prova-replicate.js
/* ============================================================
   PROVA TEMPORANEA — Replicate puo' migliorare la resa senza
   cambiare la porta?
   ------------------------------------------------------------
   Gira QUI e non in locale per un motivo preciso: le due chiavi stanno
   in Secret Manager e non devono uscirne. La piattaforma le inietta alla
   funzione; nessuno le legge, nessuno le incolla da nessuna parte.

   Cosa fa, per ogni modello chiesto:
     1. prende l'immagine di catalogo
     2. la manda a FLUX Kontext chiedendo SOLO materiale e luce
     3. rilegge il risultato con la STESSA scheda dell'accoppiatore
     4. confronta campo per campo con l'originale

   Il punto 4 e' la prova vera. Un render piu' bello che sposta un
   riquadro non e' un miglioramento: e' una porta che non vendiamo.

   DA CANCELLARE dopo la prova:
     firebase functions:delete provaReplicate --region europe-west1
   ============================================================ */

'use strict';

const fs = require('fs');
const path = require('path');
const { onRequest } = require('firebase-functions/v2/https');
const { defineSecret } = require('firebase-functions/params');
const logger = require('firebase-functions/logger');
const Anthropic = require('@anthropic-ai/sdk');

const CHIAVE_ANTHROPIC = defineSecret('ANTHROPIC_API_KEY');
const TOKEN_REPLICATE = defineSecret('REPLICATE_API_TOKEN');

// Non e' un segreto, e' un chiavistello: la funzione spende soldi veri e
// resta in piedi poche ore. Serve solo a che non la trovi un passante.
const CHIAVISTELLO = 'toscocornici-prova-2026-08';

// Scritto per non lasciare licenza creativa: solo materiale e luce, e il
// divieto esplicito di toccare il disegno. Se cambia lo stesso, la
// strada non regge.
const PROMPT = 'Photorealistic product photo of this exact interior door. '
  + 'Improve only the material and the lighting: realistic wood grain with '
  + 'visible pores and natural colour variation, soft studio lighting with '
  + 'gentle shadows, clear glass. '
  + 'Keep the door geometry absolutely identical: same number of panels, '
  + 'same panel layout and proportions, same glazing bar grid, same frame '
  + 'profile, same handle position. Do not add, remove or move any panel, '
  + 'bar or moulding. Same camera angle, same framing. Plain light background.';

const VIETATI = new Set(['minimum', 'maximum', 'exclusiveMinimum', 'exclusiveMaximum',
  'multipleOf', 'minLength', 'maxLength', 'minItems', 'maxItems', 'uniqueItems', 'pattern']);
const pulisci = (n) => {
  if (Array.isArray(n)) return n.map(pulisci);
  if (n && typeof n === 'object') {
    const out = {};
    for (const [k, v] of Object.entries(n)) {
      if (VIETATI.has(k) || k.startsWith('_')) continue;
      out[k] = pulisci(v);
    }
    return out;
  }
  return n;
};
const SCHEMA = pulisci(JSON.parse(
  fs.readFileSync(path.join(__dirname, 'dati/mappa-schema.json'), 'utf8'),
));

const ISTRUZIONI = 'Guardi l\'immagine di una porta interna e ne compili la '
  + 'scheda. Descrivi SOLO quello che vedi davvero: non dedurre, non '
  + 'completare con quello che di solito hanno le porte.\n'
  + 'Guarda con attenzione: testa (dritta o ad arco), riquadri_disposizione, '
  + 'riquadri_numero, e quanto vetro c\'e\'.';

// I campi che decidono l'accoppiamento. Se cambiano questi, l'immagine
// non rappresenta piu' il modello di catalogo.
const CHE_CONTANO = ['vetro', 'testa', 'riquadri_numero', 'riquadri_disposizione',
  'riquadri_forma', 'superficie', 'traversa', 'modanatura', 'vetro_griglia',
  'vetro_griglia_colonne', 'vetro_griglia_righe', 'vetro_lastre'];

async function kontext(token, base64jpg) {
  const r = await fetch(
    'https://api.replicate.com/v1/models/black-forest-labs/flux-kontext-pro/predictions',
    {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${token}`,
        'Content-Type': 'application/json',
        Prefer: 'wait',
      },
      body: JSON.stringify({
        input: {
          prompt: PROMPT,
          input_image: `data:image/jpeg;base64,${base64jpg}`,
          aspect_ratio: 'match_input_image',
          output_format: 'png',
          safety_tolerance: 2,
        },
      }),
    },
  );
  if (!r.ok) throw new Error(`Replicate ${r.status}: ${(await r.text()).slice(0, 300)}`);

  let p = await r.json();
  for (let i = 0; i < 60 && !['succeeded', 'failed', 'canceled'].includes(p.status); i += 1) {
    await new Promise((ok) => { setTimeout(ok, 3000); });
    p = await (await fetch(p.urls.get, { headers: { Authorization: `Bearer ${token}` } })).json();
  }
  if (p.status !== 'succeeded') throw new Error(`Replicate: ${p.error || p.status}`);

  const url = Array.isArray(p.output) ? p.output[0] : p.output;
  const img = await fetch(url);
  return Buffer.from(await img.arrayBuffer());
}

async function scheda(client, buf, tipo) {
  const r = await client.messages.create({
    model: 'claude-opus-5',
    max_tokens: 3000,
    output_config: { format: { type: 'json_schema', schema: SCHEMA } },
    messages: [{
      role: 'user',
      content: [
        { type: 'image', source: { type: 'base64', media_type: tipo, data: buf.toString('base64') } },
        { type: 'text', text: ISTRUZIONI },
      ],
    }],
  });
  if (r.stop_reason === 'refusal') return null;
  return JSON.parse(r.content.find((b) => b.type === 'text').text);
}

exports.provaReplicate = onRequest(
  {
    region: 'europe-west1',
    secrets: [CHIAVE_ANTHROPIC, TOKEN_REPLICATE],
    memory: '1GiB',
    timeoutSeconds: 540,
    maxInstances: 2,
  },
  async (req, res) => {
    if (req.query.chiave !== CHIAVISTELLO) return res.status(403).json({ errore: 'no' });

    const porte = String(req.query.porte || 'venezia,liverpool,emilia').split(',');
    const client = new Anthropic({ apiKey: CHIAVE_ANTHROPIC.value() });
    const token = TOKEN_REPLICATE.value();
    const esiti = [];

    for (const id of porte) {
      const f = path.join(__dirname, 'dati/modelli', `${id}.jpg`);
      if (!fs.existsSync(f)) { esiti.push({ id, errore: 'immagine mancante' }); continue; }
      const prima = fs.readFileSync(f);
      try {
        const dopo = await kontext(token, prima.toString('base64'));
        const a = await scheda(client, prima, 'image/jpeg');
        const b = await scheda(client, dopo, 'image/png');
        const cambiati = (a && b)
          ? CHE_CONTANO.filter((c) => a[c] !== b[c])
            .map((c) => ({ campo: c, prima: a[c], dopo: b[c] }))
          : null;
        esiti.push({
          id,
          cambiati,
          quanti: cambiati ? cambiati.length : null,
          su: CHE_CONTANO.length,
          immagine: dopo.toString('base64'),
        });
        logger.info('prova replicate', { id, cambiati: cambiati && cambiati.length });
      } catch (e) {
        esiti.push({ id, errore: e.message });
      }
    }

    return res.json({ esiti });
  },
);

File: tools/prova-replicate.py
# -*- coding: utf-8 -*-
"""
PROVA: Replicate puo' migliorare materiali e luce SENZA cambiare la porta?

Non basta guardare le immagini e dire "questa e' piu' bella". Un render
che abbellisce ma sposta un riquadro o cambia la griglia del vetro fa un
danno che l'estetica non ripaga: la scheda che mostriamo al cliente non
sarebbe piu' il prodotto.

Quindi la prova misura DUE cose:

  1. com'e' venuta      -> le immagini, da guardare
  2. se e' ancora quella porta -> si rilegge il risultato con la STESSA
     scheda che usa l'accoppiatore e si confronta campo per campo con
     l'originale. Se cambia un campo, il render ha mentito sul prodotto.

USO
    set REPLICATE_API_TOKEN=r8_...
    set ANTHROPIC_API_KEY=sk-ant-...
    python tools/prova-replicate.py

Costo: 3 immagini Kontext (~0,04 $ l'una) + 6 letture. Pochi centesimi.
"""

import base64
import io
import json
import os
import sys
import time
import urllib.error
import urllib.request

REPO = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
FUORI = os.path.join(REPO, 'tools', '_prova_replicate')
os.makedirs(FUORI, exist_ok=True)

PORTE = ['venezia', 'liverpool', 'emilia']

# Il prompt e' scritto per NON dare licenza creativa: si chiede solo il
# materiale e la luce, e si vieta esplicitamente di toccare il disegno.
# Se anche cosi' il disegno cambia, la strada non regge.
PROMPT = (
    'Photorealistic product photo of this exact interior door. '
    'Improve only the material and the lighting: realistic oak wood grain '
    'with visible pores and natural colour variation, soft studio lighting '
    'with gentle shadows, clear glass. '
    'Keep the door geometry absolutely identical: same number of panels, '
    'same panel layout and proportions, same glazing bar grid, same frame '
    'profile, same handle position. Do not add, remove or move any panel, '
    'bar or moulding. Same camera angle, same framing. Plain light '
    'background.'
)

TOKEN = os.environ.get('REPLICATE_API_TOKEN')
CHIAVE = os.environ.get('ANTHROPIC_API_KEY')
if not TOKEN:
    sys.exit('Manca REPLICATE_API_TOKEN nell\'ambiente.')
if not CHIAVE:
    sys.exit('Manca ANTHROPIC_API_KEY nell\'ambiente.')


# ---------------------------------------------------------------- Replicate

def chiama_replicate(percorso_img):
    """Manda l'immagine a Kontext e torna i byte del risultato."""
    dati = base64.b64encode(open(percorso_img, 'rb').read()).decode()
    corpo = {
        'input': {
            'prompt': PROMPT,
            'input_image': 'data:image/jpeg;base64,' + dati,
            'aspect_ratio': 'match_input_image',
            'output_format': 'png',
            'safety_tolerance': 2,
        },
    }
    req = urllib.request.Request(
        'https://api.replicate.com/v1/models/black-forest-labs/'
        'flux-kontext-pro/predictions',
        data=json.dumps(corpo).encode(),
        headers={'Authorization': 'Bearer ' + TOKEN,
                 'Content-Type': 'application/json',
                 'Prefer': 'wait'})
    try:
        r = json.load(urllib.request.urlopen(req, timeout=180))
    except urllib.error.HTTPError as e:
        sys.exit('Replicate ha risposto %s: %s' % (e.code, e.read()[:400]))

    # con Prefer: wait di solito e' gia' pronta; se no si aspetta
    for _ in range(60):
        if r.get('status') in ('succeeded', 'failed', 'canceled'):
            break
        time.sleep(3)
        r = json.load(urllib.request.urlopen(urllib.request.Request(
            r['urls']['get'], headers={'Authorization': 'Bearer ' + TOKEN})))

    if r.get('status') != 'succeeded':
        return None, r.get('error') or r.get('status')

    uscita = r['output']
    if isinstance(uscita, list):
        uscita = uscita[0]
    return urllib.request.urlopen(uscita, timeout=120).read(), None


# ---------------------------------------------------------------- la scheda

VIETATI = {'minimum', 'maximum', 'exclusiveMinimum', 'exclusiveMaximum',
           'multipleOf', 'minLength', 'maxLength', 'minItems', 'maxItems',
           'uniqueItems', 'pattern'}


def ripulisci(n):
    if isinstance(n, list):
        return [ripulisci(x) for x in n]
    if isinstance(n, dict):
        return {k: ripulisci(v) for k, v in n.items()
                if k not in VIETATI and not k.startswith('_')}
    return n


SCHEMA = ripulisci(json.load(io.open(
    os.path.join(REPO, 'tools', 'mappa-schema.json'), encoding='utf8')))

ISTRUZIONI = (
    "Guardi l'immagine di una porta interna e ne compili la scheda. "
    "Descrivi SOLO quello che vedi davvero: non dedurre, non completare "
    "con quello che di solito hanno le porte.\n"
    "Guarda con attenzione: testa (dritta o ad arco), "
    "riquadri_disposizione, riquadri_numero, e quanto vetro c'e'."
)


def leggi_scheda(byte_img, tipo):
    corpo = {
        'model': 'claude-opus-5',
        'max_tokens': 3000,
        'output_config': {'format': {'type': 'json_schema', 'schema': SCHEMA}},
        'messages': [{
            'role': 'user',
            'content': [
                {'type': 'image',
                 'source': {'type': 'base64', 'media_type': tipo,
                            'data': base64.b64encode(byte_img).decode()}},
                {'type': 'text', 'text': ISTRUZIONI},
            ],
        }],
    }
    req = urllib.request.Request(
        'https://api.anthropic.com/v1/messages',
        data=json.dumps(corpo).encode(),
        headers={'x-api-key': CHIAVE,
                 'anthropic-version': '2023-06-01',
                 'Content-Type': 'application/json'})
    r = json.load(urllib.request.urlopen(req, timeout=180))
    if r.get('stop_reason') == 'refusal':
        return None
    for b in r['content']:
        if b['type'] == 'text':
            return json.loads(b['text'])
    return None


# I campi che decidono l'accoppiamento: se cambiano questi, la porta
# mostrata non e' piu' quella del catalogo.
CHE_CONTANO = ['vetro', 'testa', 'riquadri_numero', 'riquadri_disposizione',
               'riquadri_forma', 'superficie', 'traversa', 'modanatura',
               'vetro_griglia', 'vetro_griglia_colonne', 'vetro_griglia_righe',
               'vetro_lastre']


def main():
    print('Porte da provare:', ', '.join(PORTE))
    print('')
    totale_cambi = 0
    totale_campi = 0

    for mid in PORTE:
        orig = os.path.join(REPO, 'assets', 'modelli', mid + '.webp')
        if not os.path.exists(orig):
            print(mid, '-> manca', orig)
            continue

        # Kontext vuole jpeg/png: si converte il webp
        from PIL import Image
        im = Image.open(orig).convert('RGB')
        jpg = os.path.join(FUORI, mid + '_prima.jpg')
        im.save(jpg, 'JPEG', quality=95)

        print('=== %s ===' % mid)
        print('  chiamo Replicate...')
        byte_out, errore = chiama_replicate(jpg)
        if errore:
            print('  FALLITA:', errore)
            continue
        dopo = os.path.join(FUORI, mid + '_dopo.png')
        open(dopo, 'wb').write(byte_out)
        print('  risultato scritto:', dopo)

        print('  rileggo le due schede...')
        a = leggi_scheda(open(jpg, 'rb').read(), 'image/jpeg')
        b = leggi_scheda(byte_out, 'image/png')
        if not a or not b:
            print('  lettura non riuscita')
            continue

        cambiati = []
        for c in CHE_CONTANO:
            totale_campi += 1
            if a.get(c) != b.get(c):
                cambiati.append('%s: %s -> %s' % (c, a.get(c), b.get(c)))
        totale_cambi += len(cambiati)

        if cambiati:
            print('  LA PORTA E\' CAMBIATA in %d campi su %d:'
                  % (len(cambiati), len(CHE_CONTANO)))
            for c in cambiati:
                print('     ', c)
        else:
            print('  disegno intatto: nessuno dei %d campi e\' cambiato'
                  % len(CHE_CONTANO))
        print('')

    print('=== IN TOTALE ===')
    if totale_campi:
        print('  campi cambiati: %d su %d (%.0f%%)'
              % (totale_cambi, totale_campi, 100.0 * totale_cambi / totale_campi))
    print('  immagini in:', FUORI)
    print('')
    print('  Guarda le immagini _prima/_dopo per l\'estetica,')
    print('  e i campi cambiati per sapere se e\' ancora il tuo prodotto.')


if __name__ == '__main__':
    main()
